[ChaosCenter]: Revoke the token when the user logout

namkyu1999 opened the pull request:
<!--  Thanks for sending a pull request!  -->

## Proposed changes

Issue No. #4084 

1. Add `logout` logic to Authentication server
1.1. Add Mongo Schema `revoked-token`
1.2. Add `/logout` endpoint(POST) to Authentication server
1.3. When user logout, blacklist token (createOne to 'revoked-token' collection)
2. AuthMiddleware checks whether tokens are revoked or not. (in authentication-server & graphql-server)
3. Call `/logout` in frontend when logout button clicked


## Types of changes

What types of changes does your code introduce to Litmus? Put an `x` in the boxes that apply
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices applies)

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
- [x] I have read the [CONTRIBUTING](https://github.com/litmuschaos/litmus/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the commit for DCO to be passed.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added necessary documentation (if appropriate)

## Dependency
- Please add the links to the dependent PR need to be maerged before this (if any).

## Special notes for your reviewer:

Saranya-jena commented on litmus-portal/authentication/api/main.go:
```diff
@@ -78,15 +80,27 @@ func main() {
+	}
+
+	if err = utils.CreateTTLIndex(utils.RevokedTokenCollection, db); err != nil {
+		log.Fatalf("failed to create index  %s", err)
```
instead of doing a fatal can you just log the error, would be great if you can change it for all collections creation

Saranya-jena reviewed (changes requested):
@namkyu1999 we have to do all these changes in /chaoscenter directory -  auth and web folders

Saranya-jena commented on litmus-portal/authentication/api/handlers/rest/user_handlers.go:
```diff
@@ -189,6 +192,28 @@ func LoginUser(service services.ApplicationService) gin.HandlerFunc {
+			return
+		}
+		c.JSON(200, gin.H{
+			"message": "logout succeed",
```
logout succeed -> successfully logged out

namkyu1999 commented:
@Saranya-jena 
I added logic to chaoscenter/authentication, chaoscenter/graphql, and chaoscenter/web :)

Saranya-jena reviewed (approved):
LGTM

hrishavjha reviewed (changes requested):
Hey @namkyu1999, proposed changes looks good to me. Just a small request, we've updated the REST API's setup in `chaoscenter` directory. Can you update the auth API [swagger file](https://github.com/litmuschaos/litmus/blob/master/mkdocs/docs/auth/v3.0.0/auth-api.json) and do a `yarn services` to generate frontend API and consume that. For `litmus-portal` directory everything will remain the same this change should only be inside `chaoscenter`. I'll suggest you to make changes in swagger file first and push it so that you can update the [OATS config](https://github.com/litmuschaos/litmus/blob/master/chaoscenter/web/config/oats.config.ts#L12) file with the raw link of your updated swagger file.
We can wait for @Saranya-jena's [PR](https://github.com/litmuschaos/litmus/pull/4118) to get merged first then you can update the swagger file.

namkyu1999 commented:
> Hey @namkyu1999, proposed changes looks good to me. Just a small request, we've updated the REST API's setup in `chaoscenter` directory. Can you update the auth API [swagger file](https://github.com/litmuschaos/litmus/blob/master/mkdocs/docs/auth/v3.0.0/auth-api.json) and do a `yarn services` to generate frontend API and consume that. For `litmus-portal` directory everything will remain the same this change should only be inside `chaoscenter`. I'll suggest you to make changes in swagger file first and push it so that you can update the [OATS config](https://github.com/litmuschaos/litmus/blob/master/chaoscenter/web/config/oats.config.ts#L12) file with the raw link of your updated swagger file. We can wait for @Saranya-jena's [PR](https://github.com/litmuschaos/litmus/pull/4118) to get merged first then you can update the swagger file.

Hey @hrishavjha , I synchronized the master branch code and updated the swagger file.

namkyu1999 commented:
Hey @hrishavjha , Thanks for the detailed explanation. I fixed all that you said.

One thing I want to mention in `chaoscenter/web/src/app/App.tsx`, We use a try-catch statement. But It's not a good idea because we cannot detect what's happening when errors come out. And We cannot use `forceLogout` function because We use them out of LitmusAPIProvider scope. So I deleted a try-catch statement. Is that okay?

hrishavjha reviewed (approved):
Hey @namkyu1999, following changes looks good to me 🚀

SarthakJain26 reviewed (changes requested):
Please rename the `expire_on`  field to `expires_at` in the DB and other places.

namkyu1999 commented:
> Please rename the `expire_on` field to `expires_at` in the DB and other places.

@SarthakJain26 I've included your feedback :)

amityt reviewed (approved):
LGTM
